#! /bin/sh /usr/share/dpatch/dpatch-run
## segfault1_fixed.dpatch by  <malat@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Previous segfault1.dpatch fixed a heap stack corruption, but introduced 
## DP: a regression in openjpeg, which made legal (valid) JPEG2000 impossible 
## DP: to decode. This patch properly fix the heap stack corruption while 
## DP: preserving proper decoding.
## DP: See https://bugs.debian.org/734238#53

@DPATCH@

--- openjpeg-1.3+dfsg/libopenjpeg/tcd.c	2014-04-05 14:49:32.000000000 +0200
+++ openjpeg-1.3+dfsg/libopenjpeg/tcd.c	2014-04-05 14:40:25.000000000 +0200
@@ -1407,6 +1407,15 @@
 
 	if (tcd->tcp->mct) {
 		int n = (tile->comps[0].x1 - tile->comps[0].x0) * (tile->comps[0].y1 - tile->comps[0].y0);
+
+      /* testcase 1336.pdf.asan.47.376 */
+      if ((tile->comps[0].x1 - tile->comps[0].x0) * (tile->comps[0].y1 - tile->comps[0].y0) < n ||
+        (  tile->comps[1].x1 - tile->comps[1].x0) * (tile->comps[1].y1 - tile->comps[1].y0) < n ||
+        (  tile->comps[2].x1 - tile->comps[2].x0) * (tile->comps[2].y1 - tile->comps[2].y0) < n) {
+        opj_event_msg(tcd->cinfo, EVT_ERROR, "Tiles don't all have the same dimension. Skip the MCT step.\n");
+        return false;
+      }
+
 		if (tcd->tcp->tccps[0].qmfbid == 1) {
 			mct_decode(
 					tile->comps[0].data,
